### PR TITLE
Criteria API

### PR DESCRIPTION
**Pull request by amoscatelli**
@otaviojava let's start from here !

what do you think ?

### CODE REVIEW

**Comment by genie-jnosql** [comment]
Can one of the admins verify this patch?

**Comment by otaviojava** [comment]
@amoscatelli please check the header license

**Comment by amoscatelli** [comment]
@otaviojava fixed !

**Comment by otaviojava** [comment]
@amoscatelli please update the package name to use as domain: ```org.eclipse.jnosql.mapping.mongodb```

And the documentation is important as well.
Please, don't forget to put more details in the mongodb-extension readme.

**Comment by amoscatelli** [comment]
@otaviojava I renamed the package name for the implementation, do you want me to change the package name for API too ? (jakarta.nosql)

**Comment by otaviojava** [comment]
@amoscatelli yeap, because right now it belongs to the MongoDB domain.

**Comment by amoscatelli** [comment]
@otaviojava it should be ready now.

Tell me if documentation is clear enough.
I also added 4 tests.

Next step ?

**Comment by amoscatelli** [comment]
@otaviojava done ! is this ok ?

File: mongodb-extension/pom.xml [hunk 1]
```diff
@@ -44,5 +44,17 @@
             <version>${project.version}</version>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>com.google.auto.service</groupId>
```
Comment (otaviojava) [inline mongodb-extension/pom.xml:48]: You're using reflection.
Why not use the ClassMappings?
https://github.com/eclipse/jnosql/blob/master/mapping/mapping-core/src/main/java/org/eclipse/jnosql/mapping/reflection/ClassMapping.java

You can inject the ClassMappings:

https://github.com/eclipse/jnosql/blob/master/mapping/mapping-core/src/main/java/org/eclipse/jnosql/mapping/reflection/ClassMappings.java

Comment (amoscatelli) [reply mongodb-extension/pom.xml:48]: I don't understand, are you talking about Metamodel generation ? Am I using reflection in MetamodelProcessor.
Why are you saying I am using reflection while highlighting Google Autoservice ?

https://github.com/google/auto/tree/master/service

I looked at their source code, I didn't notice reflection there.

Also, annotation processors are triggered at compile time (so they usually don't cause performance issues), are you sure I can use ClassMappings ?

Comment (otaviojava) [reply mongodb-extension/pom.xml:48]: Hum, I got it.

**Comment by otaviojava** [comment]
@amoscatelli it is good to go!
Jus fix the build, please:
https://github.com/eclipse/jnosql-mapping-extension/runs/6739631035?check_suite_focus=true

**Comment by amoscatelli** [comment]
It should be ok, it compiles on JDK 8 now

**Comment by otaviojava** [comment]
Merged thank you @amoscatelli

**Comment by otaviojava** [comment]
@amoscatelli I'll publish a snapshot of it.
Please, it is a good opportunity to write about this issue to the community.

**Comment by amoscatelli** [comment]
Thank you @otaviojava 
Exactly, what do you have in mind ?
How can I speak about this API to the community ?

**Comment by otaviojava** [comment]
Yeap, we could, if you wish, write about this feature in a blog such as Medium or Dev.to.

**Comment by amoscatelli** [comment]
That's ok to me, @otaviojava I really like the idea.

Meanwhile I also opened an issue regarding the next small/medium task.
These are the last small ones before moving to the UpdateCriteria.

https://github.com/eclipse-ee4j/nosql/issues/99

**Comment by otaviojava** [comment]
@amoscatelli 

That is easy:
1.) Create an account on https://dev.to/
2) Write about your feature
3) Publish 

:)

**Comment by amoscatelli** [comment]
@otaviojava 
Done !
https://dev.to/amoscatelli/jnosql-criteriaquery-api-preview-4l7n

**Comment by amoscatelli** [comment]
@otaviojava you said nothing about the publishing.
Was it ok ?

**Comment by otaviojava** [comment]
Yeap, it is ok.
I'll promote it today to my social medias and e-mail-list

Congratulations and well done!

**Pull request merged**